Add auto to link level speed options

edudppaz opened the pull request:
"auto" for the link level policy is missing from the choices in the ansible module. This adds the "auto" option to allow an auto negotiating interface speed to be configured on the APIC controller

akinross commented on plugins/modules/aci_interface_policy_link_level.py:
```diff
@@ -231,7 +231,7 @@ def main():
         description=dict(type="str", aliases=["descr"]),
         auto_negotiation=dict(type="bool", default="true"),
-        speed=dict(type="str", default="inherit", choices=["100M", "1G", "10G", "25G", "40G", "50G", "100G", "200G", "400G", "inherit"]),
+        speed=dict(type="str", default="inherit", choices=["100M", "1G", "10G", "25G", "40G", "50G", "100G", "200G", "400G", "auto", "inherit"]),
```
please also update documentation above to add the choice of auto

akinross replied:
could you add testcase that covers speed? I see the current test cases are not having asserts for all properties, would be great if you could have a look and increase the coverage.

edudppaz replied:
Hi @akinross , that can i fix. I take a look tonight and update this PR with the documentation and test cases.

akinross replied:
Thanks! Please re-request review when you are done because then I receive a notification.

akinross reviewed (changes requested):
thank you for the contribution, see few comments that I believe would need to be added

edudppaz commented:
@akinross  i have updated the documentation and added the new integration tests. i tried to run the whole test playbook to verify, but im getting throttled by the Cisco sandbox APIC API, and its not possible to change to certificate-based authentication. I could at least get some snippets from it:

```
Running aci_interface_policy_link_level integration test role

PLAY [aci] *********************************************************************

TASK [aci_interface_policy_link_level : Test that we have an ACI APIC host, ACI username and ACI password] ***
skipping: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify Cloud and Non-Cloud Sites in use.] ***
included: /Users/edpozo/Git/AIB/ansible_collections/cisco/aci/tests/integration/targets/aci_cloud_provider/tasks/main.yml for sandboxapicdc.cisco.com

TASK [aci_interface_policy_link_level : Test that we have an ACI APIC host, ACI username and ACI password] ***
skipping: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Set vars] ******************************
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Query system information] **************
[WARNING]: Platform darwin on host sandboxapicdc.cisco.com is using the
discovered Python interpreter at /opt/homebrew/bin/python3.11, but future
installation of another Python interpreter could change the meaning of that
path. See https://docs.ansible.com/ansible-
core/2.15/reference_appendices/interpreter_discovery.html for more information.
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Query cloud provider object] ***********
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Set vars] ******************************
skipping: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify query_cloud for all sites] ******
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Verify query_cloud for Cloud Sites] ****
skipping: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Remove Link Level Policy] **************
changed: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Add Link Level Policy (check mode)] ****
changed: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Add Link Level Policy (normal mode)] ***
changed: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify Add Link Level Policy] **********
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Add Link Level Policy again (check mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Add Link Level Policy again (normal mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify Add Link Level Policy again] ****
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Change description of Link Level Policy (check mode)] ***
changed: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Change description of Link Level Policy (normal mode)] ***
changed: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify add_policy_descr] ***************
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Change description of Link Level Policy again (check mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Change description of Link Level Policy again (normal mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify add_policy_descr_again] *********
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Add Link Level Policy again with no description (check mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Add Link Level Policy again with no description (normal mode)] ***
ok: [sandboxapicdc.cisco.com]

TASK [aci_interface_policy_link_level : Verify add_policy_again_no_descr] ******
ok: [sandboxapicdc.cisco.com] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aci_interface_policy_link_level : Add Link Level Policy with various speeds (check mode)] ***
changed: [sandboxapicdc.cisco.com] => (item=speed-100M)
changed: [sandboxapicdc.cisco.com] => (item=speed-1G)
changed: [sandboxapicdc.cisco.com] => (item=speed-10G)
changed: [sandboxapicdc.cisco.com] => (item=speed-25G)
changed: [sandboxapicdc.cisco.com] => (item=speed-40G)
changed: [sandboxapicdc.cisco.com] => (item=speed-50G)
changed: [sandboxapicdc.cisco.com] => (item=speed-100G)
changed: [sandboxapicdc.cisco.com] => (item=speed-200G)
changed: [sandboxapicdc.cisco.com] => (item=speed-400G)
changed: [sandboxapicdc.cisco.com] => (item=speed-auto)
ok: [sandboxapicdc.cisco.com] => (item=speed-inherit)

TASK [aci_interface_policy_link_level : Add Link Level Policy with various speeds (normal mode)] ***
changed: [sandboxapicdc.cisco.com] => (item=speed-100M)
changed: [sandboxapicdc.cisco.com] => (item=speed-1G)
changed: [sandboxapicdc.cisco.com] => (item=speed-10G)
changed: [sandboxapicdc.cisco.com] => (item=speed-25G)
changed: [sandboxapicdc.cisco.com] => (item=speed-40G)
changed: [sandboxapicdc.cisco.com] => (item=speed-50G)
```

edudppaz commented:
Got the whole tests running.
```
PLAY RECAP *********************************************************************
sandboxapicdc.cisco.com    : ok=43   changed=9    unreachable=0    failed=0    skipped=4    rescued=0    ignored=2   
```

akinross commented on tests/integration/targets/aci_interface_policy_link_level/tasks/main.yml:
```diff
@@ -138,6 +138,37 @@
+          speed: "{{ item }}"
+        check_mode: true
+        loop: ["100M", "1G", "10G", "25G", "40G", "50G", "100G", "200G", "400G", "auto", "inherit"]
+        register: cm_add_policy_speed
```
seems you are missing assert for the checkmode, could you add assert for proposed?

edudppaz replied:
I think i got them now, but cannot test

edudppaz replied:
"inherit" removed from check test, as is the default on ACI and is not changed .

akinross reviewed (changes requested):
i will run tests locally when you have changed the assert for checkmode

akinross commented on tests/integration/targets/aci_interface_policy_link_level/tasks/main.yml:
```diff
@@ -138,6 +138,45 @@
+        loop_control:
+          label: "speed-{{ item.0 }}"
+
+      - name: Verify each speed setting (check mode)
+        ansible.builtin.assert:
+          that:
+            - item is changed
+        loop: "{{ cm_add_policy_speed.results }}"
+        loop_control:
+          label: "speed-{{ item }}"
```
Could you add more specific asserts?

```suggestion
      - name: Verify each speed setting (check mode)
        ansible.builtin.assert:
          that:
            - "item.1.current.0.fabricHIfPol.attributes.speed == 'inherit'"
            - "item.1.proposed.fabricHIfPol.attributes.speed == item.0"
        loop: "{{ ['100M', '1G', '10G', '25G', '40G', '50G', '100G', '200G', '400G', 'auto', 'inherit'] | zip(cm_add_policy_speed.results) }}"
        loop_control:
          label: "speed-{{ item.0 }}"
```

edudppaz commented on tests/integration/targets/aci_interface_policy_link_level/tasks/main.yml:
```diff
@@ -138,6 +138,45 @@
+        cisco.aci.aci_interface_policy_link_level:
+          <<: *interface_policy_link_level_present
+          speed: "{{ item }}"
+        loop: ["100M", "1G", "10G", "25G", "40G", "50G", "100G", "200G", "400G", "auto", "inherit"]
```
Hi. I just did some changes to fetch the APIC version and depending on the version have (or not) the "auto" setting. The list of the supported settings is set based on the version. I could not use the *aci_info as it is not defined on the playbook itself (or i get an error as im doing targeted testing, not running the full suite).

Small changes on the asserts, to be more specific.



Again, could not test it locally due to throttling, but it should be ok.

akinross replied:
Not sure I understand what you mean by I cannot use *aci_info. Why not add the block on top of the test file like:

```
- name: Set vars
  ansible.builtin.set_fact:
   aci_info: &aci_info
    host: "{{ aci_hostname }}"
    port: "{{ aci_port | default(omit) }}"
    username: "{{ aci_username }}"
    password: "{{ aci_password }}"
    validate_certs: '{{ aci_validate_certs | default(false) }}'
    use_ssl: '{{ aci_use_ssl | default(true) }}'
    use_proxy: '{{ aci_use_proxy | default(true) }}'
    output_level: '{{ aci_output_level | default("info") }}'

- name: Query system information
  cisco.aci.aci_system:
    <<: *aci_info
    id: 1
    state: query
  register: version
```

Then in the task where you need define this statically you can then define `<<: *aci_info` like in the query system information above. This way you can also pass it into all other task that statically re-define the login info.

edudppaz replied:
Oh yes that can i do.. thought i just didnt want to do more "general changes" to the playbook, but is a good idea to have this available to re-use in other tests. Will fix :)

akinross reviewed (commented):
see comment response

akinross commented on tests/integration/targets/aci_interface_policy_link_level/tasks/main.yml:
```diff
@@ -138,6 +151,69 @@
+        loop: "{{ supported_speed | zip(cm_add_policy_speed.results) }}"
+        loop_control:
+          label: "speed-{{ item.0 }}"
+    ignore_errors: true
```
What is the purpose of this ignore error?

edudppaz replied:
Sorry, left behind and forgotten from testing with the throttling on sandbox. Removed :) .

akinross replied:
no worries, thanks will run the tests later and will let you know if anything still should change

akinross reviewed (approved):
LGTM

shrsr reviewed (dismissed):
LGTM

akinross reviewed (approved):
LGTM

shrsr reviewed (approved):
LGTM

samiib reviewed (approved):
LGTM

lhercot commented on plugins/modules/aci_interface_policy_link_level.py:
```diff
@@ -35,10 +35,11 @@
   speed:
     description:
-    - Determines the interface policy administrative port speed.
+    - Determines the interface policy administrative port speed. 
```
```suggestion
    - Determines the interface policy administrative port speed.
```
Please remove the trailing whitespace to help with sanity tests.

akinross reviewed (approved):
LGTM

shrsr reviewed (approved):
LGTM

samiib reviewed (approved):
LGTM

lhercot reviewed (approved):
LGTM

lhercot commented:
Thanks for your contribution!
